File: airbyte-config/config-models/src/test/java/io/airbyte/config/EnvConfigsTest.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.config;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import io.airbyte.commons.envvar.EnvVar;
import io.airbyte.commons.version.AirbyteVersion;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

@SuppressWarnings("PMD.NullAssignment")
class EnvConfigsTest {

  private Map<String, String> envMap;
  private EnvConfigs config;
  private static final String ABC = "abc";
  private static final String DEV = "dev";
  private static final String ABCDEF = "abc/def";
  private static final String ROOT = "root";
  private static final String KEY = "key";
  private static final String ONE = "one";
  private static final String TWO = "two";
  private static final String ONE_EQ_TWO = "one=two";
  private static final String NOTHING = "nothing";
  private static final String SOMETHING = "something";
  private static final String AIRBYTE = "airbyte";
  private static final String SERVER = "server";
  private static final String AIRB_SERV_SOME_NOTHING = "airbyte=server,something=nothing";
  private static final String ENV_STRING = "key=k,,;$%&^#";
  private static final String NODE_SELECTORS = ",,,";

  @BeforeEach
  void setUp() {
    envMap = new HashMap<>();
    config = new EnvConfigs(envMap);
  }

  @Test
  void ensureGetEnvBehavior() {
    assertNull(System.getenv("MY_RANDOM_VAR_1234"));
  }

  @Test
  void testAirbyteRole() {
    envMap.put(EnvVar.AIRBYTE_ROLE.name(), null);
    assertNull(config.getAirbyteRole());

    envMap.put(EnvVar.AIRBYTE_ROLE.name(), DEV);
    assertEquals(DEV, config.getAirbyteRole());
  }

  @Test
  void testAirbyteVersion() {
    envMap.put(EnvVar.AIRBYTE_VERSION.name(), null);
    assertThrows(IllegalArgumentException.class, config::getAirbyteVersion);

    envMap.put(EnvVar.AIRBYTE_VERSION.name(), DEV);
    assertEquals(new AirbyteVersion(DEV), config.getAirbyteVersion());
  }

  @Test
  void testWorkspaceRoot() {
    envMap.put(EnvVar.WORKSPACE_ROOT.name(), null);
    assertThrows(IllegalArgumentException.class, config::getWorkspaceRoot);

    envMap.put(EnvVar.WORKSPACE_ROOT.name(), ABCDEF);
    assertEquals(Paths.get(ABCDEF), config.getWorkspaceRoot());
  }

  @Test
  void testGetDatabaseUser() {
    envMap.put(EnvVar.DATABASE_USER.name(), null);
    assertThrows(IllegalArgumentException.class, config::getDatabaseUser);

    envMap.put(EnvVar.DATABASE_USER.name(), "user");
    assertEquals("user", config.getDatabaseUser());
  }

  @Test
  void testGetDatabasePassword() {
    envMap.put(EnvVar.DATABASE_PASSWORD.name(), null);
    assertThrows(IllegalArgumentException.class, config::getDatabasePassword);

    envMap.put(EnvVar.DATABASE_PASSWORD.name(), "password");
    assertEquals("password", config.getDatabasePassword());
  }

  @Test
  void testGetDatabaseUrl() {
    envMap.put(EnvVar.DATABASE_URL.name(), null);
    assertThrows(IllegalArgumentException.class, config::getDatabaseUrl);

    envMap.put(EnvVar.DATABASE_URL.name(), "url");
    assertEquals("url", config.getDatabaseUrl());
  }

  @Test
  void testworkerKubeTolerations() {
    final String airbyteServer = "airbyte-server";
    final String noSchedule = "NoSchedule";

    envMap.put(EnvVar.JOB_KUBE_TOLERATIONS.name(), null);
    assertEquals(config.getJobKubeTolerations(), List.of());

    envMap.put(EnvVar.JOB_KUBE_TOLERATIONS.name(), ";;;");
    assertEquals(config.getJobKubeTolerations(), List.of());

    envMap.put(EnvVar.JOB_KUBE_TOLERATIONS.name(), "key=k,value=v;");
    assertEquals(config.getJobKubeTolerations(), List.of());

    envMap.put(EnvVar.JOB_KUBE_TOLERATIONS.name(), "key=airbyte-server,operator=Exists,effect=NoSchedule");
    assertEquals(config.getJobKubeTolerations(), List.of(new TolerationPOJO(airbyteServer, noSchedule, null, "Exists")));

    envMap.put(EnvVar.JOB_KUBE_TOLERATIONS.name(), "key=airbyte-server,operator=Equals,value=true,effect=NoSchedule");
    assertEquals(config.getJobKubeTolerations(), List.of(new TolerationPOJO(airbyteServer, noSchedule, "true", "Equals")));

    envMap.put(EnvVar.JOB_KUBE_TOLERATIONS.name(),
        "key=airbyte-server,operator=Exists,effect=NoSchedule;key=airbyte-server,operator=Equals,value=true,effect=NoSchedule");
    assertEquals(config.getJobKubeTolerations(), List.of(
        new TolerationPOJO(airbyteServer, noSchedule, null, "Exists"),
        new TolerationPOJO(airbyteServer, noSchedule, "true", "Equals")));
  }

  @Test
  void testSplitKVPairsFromEnvString() {
    String input = "key1=value1,key2=value2";
    Map<String, String> map = config.splitKVPairsFromEnvString(input);
    assertNotNull(map);
    assertEquals(2, map.size());
    assertEquals(map, Map.of("key1", "value1", "key2", "value2"));

    input = ENV_STRING;
    map = config.splitKVPairsFromEnvString(input);
    assertNotNull(map);
    assertEquals(map, Map.of(KEY, "k"));

    input = null;
    map = config.splitKVPairsFromEnvString(input);
    assertNull(map);

    input = " key1= value1,  key2 =    value2";
    map = config.splitKVPairsFromEnvString(input);
    assertNotNull(map);
    assertEquals(map, Map.of("key1", "value1", "key2", "value2"));

    input = "key1:value1,key2:value2";
    map = config.splitKVPairsFromEnvString(input);
    assertNull(map);
  }

  @Test
  void testJobKubeNodeSelectors() {
    envMap.put(EnvVar.JOB_KUBE_NODE_SELECTORS.name(), null);
    assertNull(config.getJobKubeNodeSelectors());

    envMap.put(EnvVar.JOB_KUBE_NODE_SELECTORS.name(), NODE_SELECTORS);
    assertNull(config.getJobKubeNodeSelectors());

    envMap.put(EnvVar.JOB_KUBE_NODE_SELECTORS.name(), ENV_STRING);
    assertEquals(config.getJobKubeNodeSelectors(), Map.of(KEY, "k"));

    envMap.put(EnvVar.JOB_KUBE_NODE_SELECTORS.name(), ONE_EQ_TWO);
    assertEquals(config.getJobKubeNodeSelectors(), Map.of(ONE, TWO));

    envMap.put(EnvVar.JOB_KUBE_NODE_SELECTORS.name(), AIRB_SERV_SOME_NOTHING);
    assertEquals(config.getJobKubeNodeSelectors(), Map.of(AIRBYTE, SERVER, SOMETHING, NOTHING));
  }

}

File: airbyte-config/config-models/src/test/java/io/airbyte/config/EnvConfigsTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.config

import io.airbyte.commons.envvar.EnvVar
import io.airbyte.commons.version.AirbyteVersion
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import java.nio.file.Paths
import java.util.List
import java.util.Map

internal class EnvConfigsTest {
  private lateinit var envMap: MutableMap<String?, String?>
  private lateinit var config: EnvConfigs

  @BeforeEach
  fun setUp() {
    envMap = HashMap<String?, String?>()
    config = EnvConfigs(envMap)
  }

  @Test
  fun ensureGetEnvBehavior() {
    Assertions.assertNull(System.getenv("MY_RANDOM_VAR_1234"))
  }

  @Test
  fun testAirbyteRole() {
    envMap[EnvVar.AIRBYTE_ROLE.name] = null
    Assertions.assertNull(config.getAirbyteRole())

    envMap[EnvVar.AIRBYTE_ROLE.name] = DEV
    Assertions.assertEquals(DEV, config.getAirbyteRole())
  }

  @Test
  fun testAirbyteVersion() {
    envMap[EnvVar.AIRBYTE_VERSION.name] = null
    Assertions.assertThrows<IllegalArgumentException?>(IllegalArgumentException::class.java, Executable { config.getAirbyteVersion() })

    envMap[EnvVar.AIRBYTE_VERSION.name] = DEV
    Assertions.assertEquals(AirbyteVersion(DEV), config.getAirbyteVersion())
  }

  @Test
  fun testWorkspaceRoot() {
    envMap[EnvVar.WORKSPACE_ROOT.name] = null
    Assertions.assertThrows<IllegalArgumentException?>(IllegalArgumentException::class.java, Executable { config.getWorkspaceRoot() })

    envMap[EnvVar.WORKSPACE_ROOT.name] = ABCDEF
    Assertions.assertEquals(Paths.get(ABCDEF), config.getWorkspaceRoot())
  }

  @Test
  fun testGetDatabaseUser() {
    envMap[EnvVar.DATABASE_USER.name] = null
    Assertions.assertThrows<IllegalArgumentException?>(IllegalArgumentException::class.java, Executable { config.getDatabaseUser() })

    envMap[EnvVar.DATABASE_USER.name] = "user"
    Assertions.assertEquals("user", config.getDatabaseUser())
  }

  @Test
  fun testGetDatabasePassword() {
    envMap[EnvVar.DATABASE_PASSWORD.name] = null
    Assertions.assertThrows<IllegalArgumentException?>(IllegalArgumentException::class.java, Executable { config.getDatabasePassword() })

    envMap[EnvVar.DATABASE_PASSWORD.name] = "password"
    Assertions.assertEquals("password", config.getDatabasePassword())
  }

  @Test
  fun testGetDatabaseUrl() {
    envMap[EnvVar.DATABASE_URL.name] = null
    Assertions.assertThrows<IllegalArgumentException?>(IllegalArgumentException::class.java, Executable { config.getDatabaseUrl() })

    envMap[EnvVar.DATABASE_URL.name] = "url"
    Assertions.assertEquals("url", config.getDatabaseUrl())
  }

  @Test
  fun testworkerKubeTolerations() {
    val airbyteServer = "airbyte-server"
    val noSchedule = "NoSchedule"

    envMap[EnvVar.JOB_KUBE_TOLERATIONS.name] = null
    Assertions.assertEquals(config.getJobKubeTolerations(), mutableListOf<Any?>())

    envMap[EnvVar.JOB_KUBE_TOLERATIONS.name] = ";;;"
    Assertions.assertEquals(config.getJobKubeTolerations(), mutableListOf<Any?>())

    envMap[EnvVar.JOB_KUBE_TOLERATIONS.name] = "key=k,value=v;"
    Assertions.assertEquals(config.getJobKubeTolerations(), mutableListOf<Any?>())

    envMap[EnvVar.JOB_KUBE_TOLERATIONS.name] = "key=airbyte-server,operator=Exists,effect=NoSchedule"
    Assertions.assertEquals(config.getJobKubeTolerations(), List.of<TolerationPOJO?>(TolerationPOJO(airbyteServer, noSchedule, null, "Exists")))

    envMap[EnvVar.JOB_KUBE_TOLERATIONS.name] = "key=airbyte-server,operator=Equals,value=true,effect=NoSchedule"
    Assertions.assertEquals(
      config.getJobKubeTolerations(),
      List.of<TolerationPOJO?>(TolerationPOJO(airbyteServer, noSchedule, "true", "Equals")),
    )

    envMap[EnvVar.JOB_KUBE_TOLERATIONS.name] =
      "key=airbyte-server,operator=Exists,effect=NoSchedule;key=airbyte-server,operator=Equals,value=true,effect=NoSchedule"
    Assertions.assertEquals(
      config.getJobKubeTolerations(),
      List.of<TolerationPOJO?>(
        TolerationPOJO(airbyteServer, noSchedule, null, "Exists"),
        TolerationPOJO(airbyteServer, noSchedule, "true", "Equals"),
      ),
    )
  }

  @Test
  fun testSplitKVPairsFromEnvString() {
    var input: String? = "key1=value1,key2=value2"
    var map = config.splitKVPairsFromEnvString(input)
    Assertions.assertNotNull(map)
    Assertions.assertEquals(2, map!!.size)
    Assertions.assertEquals(map, Map.of<String?, String?>("key1", "value1", "key2", "value2"))

    input = ENV_STRING
    map = config.splitKVPairsFromEnvString(input)
    Assertions.assertNotNull(map)
    Assertions.assertEquals(map, Map.of<String?, String?>(KEY, "k"))

    input = null
    map = config.splitKVPairsFromEnvString(input)
    Assertions.assertNull(map)

    input = " key1= value1,  key2 =    value2"
    map = config.splitKVPairsFromEnvString(input)
    Assertions.assertNotNull(map)
    Assertions.assertEquals(map, Map.of<String?, String?>("key1", "value1", "key2", "value2"))

    input = "key1:value1,key2:value2"
    map = config.splitKVPairsFromEnvString(input)
    Assertions.assertNull(map)
  }

  @Test
  fun testJobKubeNodeSelectors() {
    envMap[EnvVar.JOB_KUBE_NODE_SELECTORS.name] = null
    Assertions.assertNull(config.getJobKubeNodeSelectors())

    envMap[EnvVar.JOB_KUBE_NODE_SELECTORS.name] = NODE_SELECTORS
    Assertions.assertNull(config.getJobKubeNodeSelectors())

    envMap[EnvVar.JOB_KUBE_NODE_SELECTORS.name] = ENV_STRING
    Assertions.assertEquals(config.getJobKubeNodeSelectors(), Map.of<String?, String?>(KEY, "k"))

    envMap[EnvVar.JOB_KUBE_NODE_SELECTORS.name] = ONE_EQ_TWO
    Assertions.assertEquals(config.getJobKubeNodeSelectors(), Map.of<String?, String?>(ONE, TWO))

    envMap[EnvVar.JOB_KUBE_NODE_SELECTORS.name] = AIRB_SERV_SOME_NOTHING
    Assertions.assertEquals(config.getJobKubeNodeSelectors(), Map.of<String?, String?>(AIRBYTE, SERVER, SOMETHING, NOTHING))
  }

  companion object {
    private const val ABC = "abc"
    private const val DEV = "dev"
    private const val ABCDEF = "abc/def"
    private const val ROOT = "root"
    private const val KEY = "key"
    private const val ONE = "one"
    private const val TWO = "two"
    private const val ONE_EQ_TWO = "one=two"
    private const val NOTHING = "nothing"
    private const val SOMETHING = "something"
    private const val AIRBYTE = "airbyte"
    private const val SERVER = "server"
    private const val AIRB_SERV_SOME_NOTHING = "airbyte=server,something=nothing"
    private const val ENV_STRING = "key=k,,;$%&^#"
    private const val NODE_SELECTORS = ",,,"
  }
}
